### PR TITLE
Vb 2062 configure sql for advanced timetable for establishment bni

### PR DESCRIPTION
**Pull request by AledWEvans**
## What does this pull request do?

Configures BNI for category groups and fixe errors in config for FNI to use Category Groups

## What is the intent behind these changes?

to on board new prisons correct

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/MigrateVisitService.kt [hunk 1]
```diff
@@ -244,8 +244,8 @@ class MigrateVisitService(
         getBestMatch.compare(sorted1.first(), sorted2.first())
       }
 
-      val sortedByLevelMatch = templateLocationMap.toList().sortedWith(proximityComparator)
-      return sessionLocationTemplates.find { it.reference == sortedByLevelMatch.first().first }!!
+      val bestMatchedSessionReference = templateLocationMap.toList().sortedWith(proximityComparator).first().first
+      return sessionLocationTemplates.find { it.reference == bestMatchedSessionReference }!!
     }
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/MigrateVisitService.kt:249]: Not really part of review but a slight improvement

File: src/main/resources/db/migration/V2_20__rename_visit_room.sql [hunk 2]
```diff
@@ -7,13 +7,11 @@ UPDATE legacy_data
 
 ALTER TABLE visit ALTER COLUMN visit_room drop not null;
 
-
 UPDATE visit
     SET visit_room = NULL
         FROM (SELECT v.id,v.visit_room FROM visit v JOIN legacy_data ld ON ld.visit_id = v.id) AS subquery
-    WHERE visit.id = subquery.id;
-
+    WHERE visit.id = subquery.id and visit_start < CURRENT_DATE;
```
Comment (AledWEvans) [inline src/main/resources/db/migration/V2_20__rename_visit_room.sql:13]: Not really part of ticket but found a defect in sql, this should only set visit_room for historic legacy visits

File: src/main/resources/session-template-data/session-data.csv [hunk 3]
```diff
@@ -163,16 +163,27 @@ BNI,Visits Main Hall,SOCIAL,60,5,FALSE,14:15:00,15:45:00,2023-03-10,2023-04-10,W
...
 FNI,Visits Main Hall,SOCIAL,30,4,FALSE,14:00:00,16:15:00,2023-04-18,,FRIDAY,FALSE,,FNI_NON_CAT_A_HIGH
-FNI,Visits Main Hall,SOCIAL,3,0,FALSE,14:00:00,16:00:00,2023-04-18,,FRIDAY,FALSE,,FNI_CAT_A_HIGH
+FNI,Visits CAT A,SOCIAL,3,0,FALSE,14:00:00,16:00:00,2023-04-18,,FRIDAY,FALSE,,FNI_CAT_A_HIGH
 FNI,Visits Main Hall,SOCIAL,30,4,FALSE,14:00:00,16:15:00,2023-04-18,,SATURDAY,FALSE,,FNI_NON_CAT_A_HIGH
-FNI,Visits Main Hall,SOCIAL,3,0,FALSE,14:00:00,16:00:00,2023-04-18,,SATURDAY,FALSE,,FNI_CAT_A_HIGH
+FNI,Visits CAT A,SOCIAL,3,0,FALSE,14:00:00,16:00:00,2023-04-18,,SATURDAY,FALSE,,FNI_CAT_A_HIGH
 FNI,Visits Main Hall,SOCIAL,30,4,FALSE,14:00:00,16:15:00,2023-04-18,,SUNDAY,FALSE,,FNI_NON_CAT_A_HIGH
-FNI,Visits Main Hall,SOCIAL,3,0,FALSE,14:00:00,16:00:00,2023-04-18,,SUNDAY,FALSE,,FNI_CAT_A_HIGH
\ No newline at end of file
+FNI,Visits CAT A,SOCIAL,3,0,FALSE,14:00:00,16:00:00,2023-04-18,,SUNDAY,FALSE,,FNI_CAT_A_HIGH
```
Comment (AledWEvans) [inline src/main/resources/session-template-data/session-data.csv:189]: fix up for FNI now configured for **Visits CAT A** category group

File: src/main/resources/db.scripts.mvp/R__Session_Template_Data.sql [hunk 4]
```diff
@@ -1,463 +1,483 @@
...
+    prison_code       	VARCHAR(6)    NOT NULL,
+    prison_id         	int,
+    key          		VARCHAR(50)  NOT NULL,
+    name          	    VARCHAR(100)  NOT NULL
+    );
+
+    -- Category group names are only descriptions they need to be updated when the group categories change
+        INSERT INTO tmp_session_category_group (prison_code,key,name)
+        VALUES
+            ('FNI','FNI_CAT_A_HIGH','Category A High Risk'),
```
Comment (tpmcgowan) [inline src/main/resources/db.scripts.mvp/R__Session_Template_Data.sql:415]: Given VB-1999, would be preferable for this to be "Category A (High Risk) prisoners" so we can just render it directly?

**Review by tpmcgowan — COMMENTED** [review]
Looks good - couple of minor comments

File: src/main/resources/session-template-data/session-location-data.csv [hunk 5]
```diff
@@ -6,3 +6,5 @@ BLI,BLI_G1_C2,C,2,009:010:011:012:013:014:015:016:017:018:019:020:021:022:023:02
 BLI,BLI_G1_C3,C,3,001:002:003:004:005:006:007:008:009:010:011:012:013:014:015:016:017:018:019:020:021:022:023:024,,"Wing C - Landing 3 - Cells 1 to 24"
 BLI,BLI_G2_WING,D:F,,,,Wings D and F
 BLI,BLI_G2_C2,C,2,001:002:003:004:005:006:007:008:025:026:027:028:029:030:031:032,,Wing C - Landing 2 - Cells 1 to 8 and 25 to 32
+BNI,A_WING,A,,,,"Wing A"
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/resources/session-template-data/session-location-data.csv:9]: will update

File: src/main/resources/db.scripts.mvp/R__Session_Template_Data.sql (hunk 4 above)
Comment (AledWEvans) [reply src/main/resources/db.scripts.mvp/R__Session_Template_Data.sql:415]: okay sure..

**Pull request merged**